From f741844c618e90ea56e4bdfae0c3ea1e9414627a Mon Sep 17 00:00:00 2001
From: Dimitar Dimitrov <dimitar@dinux.eu>
Date: Thu, 6 Jun 2019 09:28:06 +0300
Subject: [PATCH 10/12] Revert "Exclude r3.w0 from fixed registers"

Actually SPRUHV7C does not mark r3.w0 as neither caller-saved nor callee-saved.
So until TI clarifies this, let's mark this as fixed.

This reverts commit 4c6f177258d9f0d8313cfc8067364b7be40fa6dc.
---
 gcc/config/pru/pru.h                                           |  4 ++--
 .../gcc.target/pru/lra-framepointer-fragmentation-1.c          |  3 +--
 .../gcc.target/pru/lra-framepointer-fragmentation-2.c          | 10 ++++------
 3 files changed, 7 insertions(+), 10 deletions(-)

diff --git a/gcc/config/pru/pru.h b/gcc/config/pru/pru.h
index 7bb028eebc8..15fb637dec6 100644
--- a/gcc/config/pru/pru.h
+++ b/gcc/config/pru/pru.h
@@ -152,7 +152,7 @@
 
 #define FIXED_REGISTERS				\
   {						\
-/*   0 */  0,0,0,0, 0,0,0,0, 1,1,1,1, 0,0,1,1,	\
+/*   0 */  0,0,0,0, 0,0,0,0, 1,1,1,1, 1,1,1,1,	\
 /*   4 */  0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0,	\
 /*   8 */  0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0,	\
 /*  12 */  0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0,	\
@@ -166,7 +166,7 @@
 /* Call used == caller saved + fixed regs + args + ret vals.  */
 #define CALL_USED_REGISTERS			\
   {						\
-/*   0 */  1,1,1,1, 1,1,1,1, 1,1,1,1, 0,0,1,1,	\
+/*   0 */  1,1,1,1, 1,1,1,1, 1,1,1,1, 1,1,1,1,	\
 /*   4 */  0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0,	\
 /*   8 */  0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0,	\
 /*  12 */  0,0,0,0, 0,0,0,0, 1,1,1,1, 1,1,1,1,	\
diff --git a/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-1.c b/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-1.c
index 1432485e28f..ee1288fc2ae 100644
--- a/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-1.c
+++ b/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-1.c
@@ -12,7 +12,6 @@ uint64_t __attribute__((noinline)) test(uint64_t a, uint64_t b,
   uint64_t l1 = 0x12345678, l2 = 0x87654321, l3 = 1001, l4 = 1002;
   uint64_t l5 = 1004;
   uint32_t l6 = 2005;
-  uint16_t l7 = 2005;
   uint8_t c1 = 101, c2 = 102;
 
   /* The numerous dummy asm input operands create just
@@ -26,7 +25,7 @@ uint64_t __attribute__((noinline)) test(uint64_t a, uint64_t b,
        "r"(c), "r"(d), "r"(e), "r"(f),
        "r"(g), "r"(h), "r"(l2),
        "r"(c1), "r"(c2),
-       "r"(l3), "r"(l4), "r"(l5), "r"(l6), "r"(l7));
+       "r"(l3), "r"(l4), "r"(l5), "r"(l6));
 
   global = a+b+c+d+e+f+g+h + c1+c2 + l2;
 
diff --git a/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-2.c b/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-2.c
index 4698b4f9a85..6c98e9bf13b 100644
--- a/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-2.c
+++ b/gcc/testsuite/gcc.target/pru/lra-framepointer-fragmentation-2.c
@@ -14,7 +14,6 @@ uint64_t __attribute__((noinline)) test(uint64_t a, uint64_t b,
   uint64_t l1 = 0x12345678, l2 = 0x87654321, l3 = 1001, l4 = 1002;
   uint64_t l5 = 1004;
   uint32_t l6 = 2005;
-  uint16_t l7 = 3006;
   uint8_t c1 = 101, c2 = 102;
 
   /* The numerous dummy asm input operands create just
@@ -37,14 +36,13 @@ uint64_t __attribute__((noinline)) test(uint64_t a, uint64_t b,
        "add %0, %0, %14\n\t"
        "add %0, %0, %15\n\t"
        "add %0, %0, %16\n\t"
-       "add %0, %0, %17\n\t"
        : "=r" (l1)
        : "0" (l1), "r" (a), "r"(b),
        "r"(c), "r"(d), "r"(e), "r"(f),
        "r"(g), "r"(h), "r"(c1), "r"(c2),
-       "r"(l2), "r"(l3), "r"(l4), "r"(l5), "r"(l6), "r"(l7));
+       "r"(l2), "r"(l3), "r"(l4), "r"(l5), "r"(l6));
 
-  global = a+b+c+d+e+f+g+h + c1+c2 + l2+l3+l4+l5+l6+l7;
+  global = a+b+c+d+e+f+g+h + c1+c2 + l2+l3+l4+l5+l6;
 
   return l1;
 }
@@ -53,10 +51,10 @@ int main()
 {
   uint64_t a = test(1, 2, 3, 4, 5, 6, 7, 8);
 
-  if (a != 0x988796a6) {
+  if (a != 0x98878ae8) {
     abort();
   }
-  if (global != 0x87656362) {
+  if (global != 0x876557a4) {
     abort();
   }
   return 0;
-- 
2.11.0

